torchtune/modules/transformer.py

```diff
@@ -4,42 +4,80 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 import copy
-from typing import Optional
+from abc import abstractmethod
+from typing import Optional, Protocol, runtime_checkable
 
 import torch
 from torch import nn, Tensor
 
-from torchtune.modules import CausalSelfAttention, KVCache
+from torchtune.modules import GroupedQueryAttention, KVCache
 
 
-class TransformerDecoderLayer(nn.Module):
+@runtime_checkable
+class TransformerLayer(Protocol):
+    """Transformer layer meant to by used in TransformerDecoder"""
+
+    @abstractmethod
+    def forward(
+        self,
+        x: Tensor,
+        *,
+        mask: Optional[Tensor] = None,
+        encoder_input: Optional[Tensor] = None,
+        encoder_mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+    ) -> Tensor:
+        """
+        Args:
+            x (Tensor): input sequence tensor
+            mask (Optional[Tensor]): boolean tensor which contains the attention mask
+            encoder_input (Optional[Tensor]): secondary input sequence tensor
+                transformer layers. This input is ignored and does nothing in self attention.
+            encoder_mask (Optional[Tensor]):  boolearn tensor to mask connections between x and encoder_input
+            input_pos (Optional[Tensor]): Optional tensor which contains the position ids of each token.
+
+        Returns:
+            Tensor: output tensor with same shape as input
+        """
+
+
+class TransformerSelfAttentionLayer(nn.Module):
     """Transformer layer derived from the Llama2 model. Normalization is applied before the attention **and** FF layer.
 
     Args:
-        attn (CausalSelfAttention): Attention module.
+        attn (GroupedQueryAttention): Attention module.
         mlp (nn.Module): Feed-forward module.
-        sa_norm (nn.Module): Normalization to be applied before self-attention.
-        mlp_norm (nn.Module): Normalization to be applied before the feed-forward layer.
+        attn_norm (Optinoal[nn.Module]): Normalization to be applied before self-attention.
+        mlp_norm (Optional[nn.Module]): Normalization to be applied before the feed-forward layer.
+        attn_scale (Optinoal[nn.Module]): Module to scale self-attention output.
+        mlp_scale (Optional[nn.Module]): Module to scale the feed-forward output.
     """
 
     def __init__(
         self,
-        attn: CausalSelfAttention,
+        attn: GroupedQueryAttention,
         mlp: nn.Module,
-        sa_norm: nn.Module,
-        mlp_norm: nn.Module,
+        *,
+        attn_norm: Optional[nn.Module] = None,
+        mlp_norm: Optional[nn.Module] = None,
+        attn_scale: Optional[nn.Module] = None,
+        mlp_scale: Optional[nn.Module] = None,
```

Comment:
I agree with this in spirit, but wouldn't we be instantiating the `nn.Identity()` in the function definition? Also the `nn.Identity` may then be shared across multiple usages of `TransformerSelfAttentionLayer`. While is' not as bad as e.g. a module that actually has trainable params, I can still see this being troublesome.

Reply:
Don't we already do this for Linear vs LoRALinear? Or how is this different?

Reply:
Example: https://github.com/pytorch/torchtune/blob/main/torchtune/models/llama3/_component_builders.py#L90

Reply:
Oh my comment is specifically about the usage as a default argument on init. Like

```
class ModuleWithDefaults(nn.Module):
	def __init__(
		self,
		submodule: nn.Module = nn.Identity(),
	):
		super().__init__()
		self.submodule = submodule
```

is different than

```
class ModuleWithoutDefaults(nn.Module):
	def __init__(
		self,
		submodule: nn.Module,
	):
		super().__init__()
		self.submodule = submodule

def build_module_without_defaults():
	return ModuleWithoutDefaults(nn.Identity())
```

because in the first case we actually create `nn.Identity()` as soon as the class is defined, which only happens once. But in the second case we only create it when we're actually creating a specific instance of `ModuleWithoutDefaults`. So the second case will have different `nn.Identity`s created each time `build_module_without_defaults` is called. The first one will only have a single one used for every subsequent call `ModuleWithDefaults()` (basically the same reason why we don't want to use a dict as a default argument in class init).

Reply:
But yeah what we could do is just define the `nn.Identity()` in the builder and pass it in from there (maybe this is what you were suggesting anyways based on the LoRA code pointer?)